magma/circuit.py

```diff
@@ -830,3 +830,52 @@ def wrapped(*args, **kwargs):
 def DeclareCoreirCircuit(*args, **kwargs):
     return DeclareCircuit(*args, **kwargs,
                           renamed_ports=coreir_port_mapping)
+
+
+class _CircuitBuilderMeta(type):
+    pass
+
+
+def builder_method(func):
+
+    @wraps(func)
+    def _wrapped(this, *args, **kwargs):
+        with _DefinitionContextManager(this._context):
+            result = func(this, *args, **kwargs)
+        return result
+
+    return _wrapped
+
+
+class CircuitBuilder(metaclass=_CircuitBuilderMeta):
+    def __init__(self, name):
+        try:
+            context = _definition_context_stack.peek()
+            context.add_builder(self)
+        except IndexError:  # builders must be inside a definition context
+            raise Exception("Can not instance a circuit builder outside a "
+                            "definition")
+        self._name = name
+        self._io = SingletonInstanceIO()
+        self._finalized = False
+        self._context = DefinitionContext(StagedPlacer(self._name))
+
+    def _port(self, name):
+        return self._io.ports[name]
+
+    def _add_port(self, name, typ):
+        self._io.add(name, typ)
+        setattr(self, name, self._io.inst_ports[name])
+
+    def _finalize(self):
+        pass
+
+    def finalize(self):
+        if self._finalized:
+            raise Exception("Can only call finalize on a CircuitBuilder once")
+        self._finalize()
+        bases = (AnonymousCircuitType,)
+        dct = {"io": self._io, "_context_": self._context, "name": self._name}
```

Comment:
I wonder if we want to allow the builder to passthrough other names into the `dct` context?  Magma relies on a a lot of special names in the class namespace for certain features.  Alternatively, we should consider a different API for these features.

Reply:
Good idea. Added a hook to do that called `set_namespace_key`